### PR TITLE
Auto-Differentiation Cost Class Implementation

### PR DESCRIPTION
**Pull request by gavin-hyl**
- **Unfinished, problems with CodeGen**
- **Evaluate and gradient now workingl**
- **Completed autodiff class and basic tests.**
- **Fix autodiff class identifier problem. Add related tests.**
The tests (as of now) are mostly from the user-defined cost class' tests. I feel like implementing the consistency tests will be much easier after merging the user-defined cost branch, since we can define functions and their "correct derivatives" in a much more streamlined way.

### CODE REVIEW

**Comment by Zolkin1** [comment]
@gavin-hyl It's looking good, but I had a thought last night. I think we should make a separate AutoDiff wrapper class that just handles all the interface with the autodiff/codegen. Then, we can re-use this wrapper for the constraints and models as necessary. 

How hard would that be to do? We can make separate top-level folder with an associated Cmake file to hold this code.

File: functions/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+message(STATUS "Building costs")
+
+set(FN_HEADER_INC_DIR "${torc_SOURCE_DIR}/functions/include")
+
+set(FN_HEADER_LIST
+#    "${FN_HEADER_INC_DIR}/linear_fn.h"
```
Comment (Zolkin1) [inline functions/CMakeLists.txt]: Shouldn't these be uncommented?

Comment (gavin-hyl) [reply functions/CMakeLists.txt]: Yes...apologies for the oversight.

**Review by Zolkin1 — APPROVED** [review]
lgtm

**Pull request merged**